### PR TITLE
fix(docutils): better parsing/updating of nav tree in `mkdocs.yml`

### PR DESCRIPTION
**Pull request by boneskull**
After we build the reference docs, we need to update the `nav` prop of the given `mkdocs.yml` before building the site.  This was only _sorta_ working before; it did not take into account the myriad ways in which the data structure could be expressed (particularly, it did not understand "custom names").

I think I've done this in such a way that if a custom name is provided (they must be provided manually by hand-editing `mkdocs.yml`), it retains them _unless_ the file in question disappears.  Or that's the idea. Can't really be too sure; needs tests.

This change makes an attempt to parse the `nav` prop into something "normalized", then the data is processed and recomplexified before writing out to `mkdocs.yml`.  However, I've disabled the ability to define a custom header for command docs and/or omit the header entirely, as the latter especially was causing extra complexity and it's already bad enough.  I think we can re-enable "custom header" somehow, if needed.

Also added an `--all` flag which causes the nav to be updated with _all_ the TypeDoc-generated
content--not just command docs.

### CODE REVIEW

**Comment by boneskull** [comment]
Current dependencies on/for this PR:
* master
  * **PR #18153** <a href="https://app.graphite.dev/github/pr/appium/appium/18153" target="_blank">Graphite</a> 
    * **PR #18154** <a href="https://app.graphite.dev/github/pr/appium/appium/18154" target="_blank">Graphite</a> 
      * **PR #18155** <a href="https://app.graphite.dev/github/pr/appium/appium/18155" target="_blank">Graphite</a> 
        * **PR #18156** <a href="https://app.graphite.dev/github/pr/appium/appium/18156" target="_blank">Graphite</a> 
          * **PR #18157** <a href="https://app.graphite.dev/github/pr/appium/appium/18157" target="_blank">Graphite</a> 
            * **PR #18158** <a href="https://app.graphite.dev/github/pr/appium/appium/18158" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18158?utm_source=stack-comment).

**Comment by boneskull** [comment]
Current dependencies on/for this PR:
* master
  * **PR #18153** <a href="https://app.graphite.dev/github/pr/appium/appium/18153" target="_blank">Graphite</a> 
    * **PR #18154** <a href="https://app.graphite.dev/github/pr/appium/appium/18154" target="_blank">Graphite</a> 
      * **PR #18155** <a href="https://app.graphite.dev/github/pr/appium/appium/18155" target="_blank">Graphite</a> 
        * **PR #18156** <a href="https://app.graphite.dev/github/pr/appium/appium/18156" target="_blank">Graphite</a> 
          * **PR #18157** <a href="https://app.graphite.dev/github/pr/appium/appium/18157" target="_blank">Graphite</a> 
            * **PR #18158** <a href="https://app.graphite.dev/github/pr/appium/appium/18158" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18158?utm_source=stack-comment).

File: packages/docutils/lib/builder/nav.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+ * a `name` property)
+ */
+export function parseNav(nav: MkDocsYmlNav): ParsedNavData[] {
+  let parsedNav: ParsedNavData[] = [];
+  const entries = Object.entries(nav);
+  type QueueItem = {
+    entries: typeof entries;
+    keypath: string;
+  };
+  const queue: QueueItem[] = [{entries: Object.entries(nav), keypath: ''}];
```
Comment (jlipps) [inline packages/docutils/lib/builder/nav.ts:75]: why not `{entries, keypath: ''}`

**Review by jlipps — APPROVED** [review]

File: packages/docutils/lib/builder/nav.ts (hunk 1 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:75]: good catch

File: packages/docutils/lib/builder/nav.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    for (const [key, item] of entries) {
+      if (_.isString(item)) {
+        const navData: ParsedNavData = {
+          keypath: keypath ? `${keypath}.${key}` : key,
+          fileOrUrl: item,
+        };
+
+        // if the key is not convertible to a number, it's a name
+        // which was manually put there by somebody.
+        if (isNaN(Number(key))) {
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:88]: can we use Number.isNaN?

File: packages/docutils/lib/builder/nav.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+          navData.name = key;
+        }
+        parsedNav = [...parsedNav, navData];
+      } else if (_.isObject(item)) {
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:92]: did you mean isPlainObject?

File: packages/docutils/lib/builder/nav.ts [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+/**
+ * Finds all items within the list of parsed nav data which correpsond to the header.  This is
+ * imperfect, as it's possible for the header string to appear in multiple places in the nav, but let's just ignore that until we can't.
+ * @param navData Some parsed nav data
+ * @param header Header name
+ */
+function itemsForHeader(navData: ParsedNavData[], header: string) {
+  return _.filter(navData, (item) => _.toPath(item.keypath).includes(header));
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:108]: can keypath be null here?

File: packages/docutils/lib/builder/nav.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+export async function updateNav({
+  cwd = process.cwd(),
+  mkdocsYml: mkDocsYmlPath,
+  typedocJson: typeDocJsonPath,
+  all = false,
+  dryRun = false,
+}: UpdateNavOpts = {}) {
+  // we need `mkdocs.yml` to update
+  // and we need `typedoc.json` to know where to look for the command docs
+  [mkDocsYmlPath, typeDocJsonPath] = await Promise.all([
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:129]: I prefer to use B.all, it's more flexible to cases where the input size is zero

File: packages/docutils/lib/builder/nav.ts [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  // we don't need the `name` prop, since the name is already present in the keypath.
+  const newData: Omit<ParsedNavData, 'name'>[] = [];
+
+  for (const dir of dirs) {
+    const newDataForDir: Omit<ParsedNavData, 'name'>[] = [];
+    const newRefFilepaths = await findRelativeMarkdownFiles(dir, mkDocsDocsDir);
+
+    const header = all ? _.startCase(path.basename(dir)) : DEFAULT_NAV_HEADER;
+
+    const headerItems = itemsForHeader(navData, header);
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:200]: I usually use verbs for method or function names

File: packages/docutils/lib/builder/nav.ts [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    const yaml = stringifyYaml(mkDocsYml);
+    log.debug('Writing to %s:\n%s', mkDocsYmlPath, yaml);
+    await safeWriteFile(mkDocsYmlPath, yaml, true);
+    log.success(
+      'Updated MkDocs navigation config for reference docs; please run "git add -A %s" and commit this change',
+      relMkDocsYmlPath
+    );
+  } else {
+    log.info('No changes needed for MkDocs config at %s', relMkDocsYmlPath);
+  }
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:266]: this method is quite long and hard to understand. Maybe it makes sense to extract some code from it?

File: packages/docutils/lib/builder/nav.ts [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+      // we append indices or names to this keypath
+      const rootHeaderKeypath = rootHeaderKeypathParts
+        .slice(0, rootHeaderKeypathParts.indexOf(header) + 1)
+        .join('.');
+
+      for (const [offset, newRefFilepath] of newRefFilepaths.entries()) {
+        const data = headerItems.find((item) => item.fileOrUrl === newRefFilepath);
+        if (data?.name) {
+          newDataForDir.push({
+            keypath: `${rootHeaderKeypath}.${offset}.${data.name}`,
```
Comment (mykola-mokhnach) [inline packages/docutils/lib/builder/nav.ts:219]: the condition could be simplified to `keypath: `${rootHeaderKeypath}.${offset}${data.name ? ('.' + data.name) : ''}`,`

File: packages/docutils/lib/builder/nav.ts (hunk 2 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:88]: why?

File: packages/docutils/lib/builder/nav.ts (hunk 3 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:92]: no, it doesn't need to be a plain object

File: packages/docutils/lib/builder/nav.ts (hunk 4 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:108]: no, `keypath` cannot be null; see the type of `ParsedNavData`

File: packages/docutils/lib/builder/nav.ts (hunk 5 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:129]: that would mean introducing bluebird where I don't necessarily need it though.  the input size is known

File: packages/docutils/lib/builder/nav.ts (hunk 6 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:200]: okay, that's probably better, I will change it

File: packages/docutils/lib/builder/nav.ts (hunk 8 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:219]: I didn't want to embed a ternary in a template, I guess.

File: packages/docutils/lib/builder/nav.ts (hunk 3 above)
Comment (mykola-mokhnach) [reply packages/docutils/lib/builder/nav.ts:92]: is it still possible/makes sense to call Object.entries on a non-plain object?

File: packages/docutils/lib/builder/nav.ts (hunk 2 above)
Comment (mykola-mokhnach) [reply packages/docutils/lib/builder/nav.ts:88]: https://www.w3schools.com/jsref/jsref_isnan_number.asp

isNaN does unnecessary implicit conversion while the latter does not

Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:88]: fair enough

File: packages/docutils/lib/builder/nav.ts (hunk 7 above)
Comment (boneskull) [reply packages/docutils/lib/builder/nav.ts:334]: @mykola-mokhnach OK, I've pushed changes based on your review

**Pull request merged**